File: windows/Marquee.Core.Tests/RequestAllMissingTests.cs
using System.Text.Json.Nodes;
using Marquee.Core.Api;
using Marquee.Core.Models;
using Marquee.Core.Tests.Support;

namespace Marquee.Core.Tests;

// A household member's "Request all N missing" on a franchise row, as the
// Mac's RequestAllMissingTests: the optional franchise.requestAllMissing
// (missing from an older server hides the button), the button's words, the
// result, and the one POST it sends.

public sealed class RequestAllMissingTests
{
    private static readonly Uri Base = new("http://127.0.0.1:3000");

    /// <summary>The doc's title detail with its franchise edited.</summary>
    private static TitleDetail Detail(Action<JsonObject> edit)
    {
        var json = JsonNode.Parse(Fixtures.Read("title-detail"))!.AsObject();
        edit(json["franchise"]!.AsObject());
        return Json.Decode<TitleDetail>(json.ToJsonString());
    }

    [Fact]
    public void DecodesWithAndWithoutTheField()
    {
        var member = Detail(franchise =>
        {
            franchise["addAllMissing"] = new JsonArray();
            franchise["requestAllMissing"] = new JsonArray(
                new JsonObject { ["mediaType"] = "movie", ["tmdbId"] = 57800 },
                new JsonObject { ["mediaType"] = "movie", ["tmdbId"] = 278154 });
        });
        Assert.Equal([new TitleId(MediaType.Movie, 57800), new TitleId(MediaType.Movie, 278154)], member.Franchise!.RequestAllMissing);
        Assert.Equal(2, member.Franchise.RequestAllCount);

        var older = Detail(franchise => franchise.Remove("requestAllMissing"));
        Assert.Null(older.Franchise!.RequestAllMissing);
        Assert.Equal(0, older.Franchise.RequestAllCount);

        // The doc's example is the admin's view: nothing to request.
        var admin = Fixtures.Decode<TitleDetail>("title-detail");
        Assert.Empty(admin.Franchise!.RequestAllMissing!);
        Assert.Equal(0, admin.Franchise.RequestAllCount);
    }

    [Fact]
    public void ButtonWords()
    {
        Assert.Equal("Request all 4 missing", TitleFranchise.RequestAllLabel(4, busy: false));
        Assert.Equal("Requesting…", TitleFranchise.RequestAllLabel(4, busy: true));
        Assert.Equal("Request all 4 missing titles?", TitleFranchise.RequestAllConfirmation(4));
        Assert.Equal("Request all 1 missing title?", TitleFranchise.RequestAllConfirmation(1));
    }

    [Fact]
    public void ResultDecodes()
    {
        var result = Fixtures.Decode<RequestAllMissingResult>("request-all-missing");
        Assert.Equal(4, result.Total);
        Assert.Equal(2, result.Requested);
        Assert.Equal([57800, 278154], result.Refused.Select(refusal => refusal.TmdbId));
        Assert.Equal(MediaType.Movie, result.Refused[0].MediaType);
        Assert.StartsWith("Requested 2 of 4.", result.Message, StringComparison.Ordinal);
    }

    [Fact]
    public async Task SendsOnePostWithNoBodyAndRecordsTheChange()
    {
        var stub = new StubHttpMessageHandler();
        stub.AnswerFixture("request-all-missing");
        var events = new ServerEvents();
        var api = new MarqueeApi(new ApiClient(Base, "mqt_testtesttesttesttesttesttesttesttesttesttes", stub), events);

        var result = await api.Titles.RequestAllMissingAsync(MediaType.Movie, 425);

        var request = Assert.Single(stub.Requests);
        Assert.Equal("POST", request.Method.Method);
        Assert.Equal("/api/v1/titles/movie/425/request-all-missing", request.Path);
        Assert.Equal("", request.Body);
        Assert.Equal(1, events.Revision(ServerChange.Requests));
        Assert.Equal(1, events.Revision(ServerChange.Library));
        Assert.Equal(2, result.Requested);
    }

    [Fact]
    public async Task TheAdminIsTurnedAway()
    {
        var stub = new StubHttpMessageHandler();
        stub.AnswerJson(403, """{"error":"The admin adds titles straight to the library — use Add all.","code":"forbidden"}""");
        var api = new MarqueeApi(new ApiClient(Base, "mqt_testtesttesttesttesttesttesttesttesttesttes", stub));

        var error = await Assert.ThrowsAsync<ApiException>(() => api.Titles.RequestAllMissingAsync(MediaType.Movie, 425));
        Assert.Equal(ApiErrorKind.Forbidden, error.Kind);
    }
}

File: mac/MarqueeTests/RequestAllMissingTests.swift
import XCTest
@testable import Marquee

// A household member's "Request all N missing" on a franchise row: the
// optional `franchise.requestAllMissing` (missing from an older server hides
// the button), the result, and the title page's model sending it and
// reloading the page afterwards.

@MainActor
final class RequestAllMissingTests: XCTestCase {
    override func tearDown() {
        StubURLProtocol.handler = nil
        StubURLProtocol.requests = []
        super.tearDown()
    }

    private func fixture(_ name: String) throws -> [String: Any] {
        let url = Bundle(for: Self.self).resourceURL!.appendingPathComponent("Fixtures/api/\(name).json")
        return try XCTUnwrap(JSONSerialization.jsonObject(with: Data(contentsOf: url)) as? [String: Any])
    }

    /// The doc's title detail with its franchise edited.
    private func detailJSON(_ edit: (inout [String: Any]) -> Void) throws -> Data {
        var json = try fixture("title-detail")
        var franchise = try XCTUnwrap(json["franchise"] as? [String: Any])
        edit(&franchise)
        json["franchise"] = franchise
        return try JSONSerialization.data(withJSONObject: json)
    }

    private func memberDetail() throws -> Data {
        try detailJSON {
            $0["addAllMissing"] = []
            $0["requestAllMissing"] = [["mediaType": "movie", "tmdbId": 57800], ["mediaType": "movie", "tmdbId": 278154]]
        }
    }

    func testDecodesWithAndWithoutTheField() throws {
        let member = try APIClient.decoder.decode(API.TitleDetail.self, from: memberDetail())
        XCTAssertEqual(member.franchise?.requestAllMissing, [API.TitleID(.movie, 57800), API.TitleID(.movie, 278154)])

        let older = try APIClient.decoder.decode(API.TitleDetail.self, from: detailJSON { $0.removeValue(forKey: "requestAllMissing") })
        XCTAssertNil(older.franchise?.requestAllMissing, "An older server leaves it out: no button")
    }

    func testResultDecodes() throws {
        let data = try JSONSerialization.data(withJSONObject: fixture("request-all-missing"))
        let result = try APIClient.decoder.decode(API.RequestAllMissingResult.self, from: data)
        XCTAssertEqual(result.total, 4)
        XCTAssertEqual(result.requested, 2)
        XCTAssertEqual(result.refused.map(\.tmdbId), [57800, 278154])
        XCTAssertTrue(result.message.hasPrefix("Requested 2 of 4."))
    }

    func testRequestAllSendsOnePostThenReloadsThePage() async throws {
        let detail = try memberDetail()
        let result = try JSONSerialization.data(withJSONObject: fixture("request-all-missing"))
        let client = APIClient(baseURL: URL(string: "http://127.0.0.1:3000")!, token: "mqt_test", session: StubURLProtocol.session())
        let api = MarqueeAPI(client: client, events: ServerEvents())
        StubURLProtocol.handler = { request in
            let body = request.httpMethod == "POST" ? result : detail
            return (200, StubURLProtocol.apiHeaders, body)
        }

        let screen = TitleDetailModel(id: API.TitleID(.movie, 425))
        await screen.load(api)
        XCTAssertEqual(screen.detail?.franchise?.requestAllMissing?.count, 2)
        StubURLProtocol.requests = []

        await screen.requestAllMissing()

        XCTAssertEqual(StubURLProtocol.requests.map { "\($0.httpMethod ?? "") \($0.url?.path ?? "")" }, [
            "POST /api/v1/titles/movie/425/request-all-missing",
            "GET /api/v1/titles/movie/425",
        ])
        XCTAssertEqual(
            screen.requestAllResult,
            "Requested 2 of 4. You've used your 2 movie requests for a week. You can ask again in 7 days."
        )
        XCTAssertFalse(screen.isRequestingAll)
    }

    func testNothingToRequestSendsNothing() async throws {
        let detail = try detailJSON { $0["requestAllMissing"] = [] }
        let client = APIClient(baseURL: URL(string: "http://127.0.0.1:3000")!, token: "mqt_test", session: StubURLProtocol.session())
        let api = MarqueeAPI(client: client, events: ServerEvents())
        StubURLProtocol.handler = { _ in (200, StubURLProtocol.apiHeaders, detail) }

        let screen = TitleDetailModel(id: API.TitleID(.movie, 603))
        await screen.load(api)
        StubURLProtocol.requests = []
        await screen.requestAllMissing()
        XCTAssertTrue(StubURLProtocol.requests.isEmpty)
        XCTAssertNil(screen.requestAllResult)
    }

    func testAFailureShowsTheServersMessage() async throws {
        let detail = try memberDetail()
        let client = APIClient(baseURL: URL(string: "http://127.0.0.1:3000")!, token: "mqt_test", session: StubURLProtocol.session())
        let api = MarqueeAPI(client: client, events: ServerEvents())
        StubURLProtocol.handler = { request in
            request.httpMethod == "POST"
                ? StubURLProtocol.json(502, #"{"error":"Couldn't look this title up with TMDb right now.","code":"upstream"}"#)
                : (200, StubURLProtocol.apiHeaders, detail)
        }
        let screen = TitleDetailModel(id: API.TitleID(.movie, 425))
        await screen.load(api)
        await screen.requestAllMissing()
        XCTAssertEqual(screen.requestAllResult, "Couldn't look this title up with TMDb right now.")
        XCTAssertFalse(screen.isRequestingAll)
    }
}
